crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3141,9 +3169,10 @@ impl<'db> TypeInferenceBuilder<'db> {
             // TODO: async iterables/iterators! -- Alex
             todo_type!("async iterables/iterators")
         } else {
-            iterable_ty
-                .iterate(self.db())
-                .unwrap_with_diagnostic(&self.context, iterable.into())
+            iterable_ty.try_iterate(self.db()).unwrap_or_else(|err| {
+                err.report_diagnostic(&self.context, iterable.into());
+                err.fallback_element_type()
```

Comment:
This is a common pattern and it would be nice if it could be a bit less verbose, but I guess there's no obvious way to do that short of a custom `Result` like type.

Reply:
I had `fallback_element_type_with_diagnostic` but it just felt sooo long. I don't mind the repetitiveness here. It's simple and more flexible, e.g. you don't have to call `fallback_element_type` if you only want to report a diagnostic. 

Overall, I don't think we should spend much time on this: While repetitive, the call sites are very limited.